### PR TITLE
Add initial draft of ML overview document

### PR DESCRIPTION
**Pull request by mrocklin**
See https://github.com/dask/dask/issues/11095

cc @jrbourbeau @scharlottej13 

This isn't yet done.  Some TODOs:

- [x] Make better example for Optuna
- [x] Make better example for XGBoost
- [x] Add more example links throughout

But I figured I'd pause here and ask for large scale feedback

### CODE REVIEW

**Comment by mrocklin** [comment]
I also added a video, but maybe that's a distraction.  Mostly I used it because I saw it in the Optuna docs

**Comment by mrocklin** [comment]
TODOs are done.  I think that this is maybe ok.  @scharlottej13 if you're around and can take a look that would be welcome.  I'm hopeful that this will help with dask.org links.

File: docs/source/ml.rst [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+   client = cluster.get_client()
+
+   futures = [
+       client.submit(study.optimize, objective, n_trials=1, pure=False) for _ in range(500)
+   ]
+   wait(futures)
+
+   print(study.best_params)
+
+For a more fully worked example see :bdg-link-primary:`this Optuna+XGBoost example <https://docs.coiled.io/user_guide/usage/dask/hpo.html>`.
```
Comment (scharlottej13) [inline docs/source/ml.rst:69]:
```suggestion
For a more fully worked example see :bdg-link-primary:`this Optuna+XGBoost example <https://docs.coiled.io/user_guide/usage/dask/hpo.html?utm_source=dask-docs&utm_medium=ml>`.
```

File: docs/source/ml.rst [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+.. code-block:: python
+
+   from dask.distributed import LocalCluster
+
+   cluster = LocalCluster(processes=False)  # replace this with some scalable cluster
+   client = cluster.get_client()
+
+   def train_and_score(params: dict) -> float:
+       # TODO: your code here
```
Comment (scharlottej13) [inline docs/source/ml.rst:88]:
```suggestion
```

File: docs/source/ml.rst [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+Popular GBT libraries have native Dask support which allows you to train models
+on very large datasets in parallel.  Both XGBoost and LightGBM have pretty good
+documentation and examples:
+
+-  `XGBoost <https://xgboost.readthedocs.io/en/stable/tutorials/dask.html>`_
+-  `LightGBM <https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#dask>`_
+
+For convenience, here is a copy-pastable example using Dask Dataframe, XGBoost,
+and the Dask LocalCluster to train on randomly generated data
```
Comment (scharlottej13) [inline docs/source/ml.rst:117]:
```suggestion
and the Dask LocalCluster to train on randomly generated data:
```

File: docs/source/ml.rst [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+           d_train = xgboost.dask.DaskDMatrix(client, X_train, y_train, enable_categorical=True)
+           model = xgboost.dask.train(
+               ...
+               d_train,
+           )
+           predictions = xgboost.dask.predict(client, model, X_test)
+
+           score = ...
+
+For a more fully worked example see :bdg-link-primary:`this XGBoost example <https://docs.coiled.io/user_guide/usage/dask/xgboost.html>`.
```
Comment (scharlottej13) [inline docs/source/ml.rst:142]:
```suggestion
For a more fully worked example see :bdg-link-primary:`this XGBoost example <https://docs.coiled.io/user_guide/usage/dask/xgboost.html?utm_source=dask-docs&utm_medium=ml>`.
```

File: docs/source/ml.rst [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+For a more fully worked example see :bdg-link-primary:`Batch Scoring for Computer Vision Workloads (video) <https://developer.download.nvidia.com/video/gputechconf/gtc/2019/video/S9198/s9198-dask-and-v100s-for-fast-distributed-batch-scoring-of-computer-vision-workloads.mp4>`.
+
+Batch Prediction with Dask Dataframe
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes we access our that we want to process with our model with a higher
+level Dask API, like Dask Dataframe or Dask Array.  This is more common with
+record data, for example if we had a set of patient records and we wanted to
+see which were likely to become ill
```
Comment (scharlottej13) [inline docs/source/ml.rst:192]:
```suggestion
see which patients were likely to become ill
```

File: docs/source/ml.rst [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+   import dask.dataframe as dd
+
+   df = dd.read_parquet("/path/to/my/data.parquet")
+
+   model = load_model("/path/to/my/model")
+
+   # pandas code
+   # predictions = model.predict(df)
+   predictions.to_parquet("/path/to/results.parquet")
```
Comment (scharlottej13) [inline docs/source/ml.rst:204]:
```suggestion
   # predictions.to_parquet("/path/to/results.parquet")
```
to show the difference between pandas and dask dataframe syntax?

File: docs/source/ml.rst [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+   model = client.submit(load_model, path_to_model)
+   predictions = client.map(predict, filenames, model=model)
+   results = client.gather(predictions)
+
+For a more fully worked example see :bdg-link-primary:`Batch Scoring for Computer Vision Workloads (video) <https://developer.download.nvidia.com/video/gputechconf/gtc/2019/video/S9198/s9198-dask-and-v100s-for-fast-distributed-batch-scoring-of-computer-vision-workloads.mp4>`.
+
+Batch Prediction with Dask Dataframe
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes we access our that we want to process with our model with a higher
```
Comment (scharlottej13) [inline docs/source/ml.rst:189]:
```suggestion
Sometimes we want to process with our model with a higher
```

**Comment by scharlottej13** [comment]
(could be for a future iteration) I could also see including a joblib + scikit-learn section. Maybe there are also utility functions from dask-ml that people commonly use?

**Comment by mrocklin** [comment]
> (could be for a future iteration) I could also see including a joblib + scikit-learn section. Maybe there are also utility functions from dask-ml that people commonly use?

I don't really see this all that often (xgboost and pytorch seem to be more common with dask workloads).  I could be wrong though.  I have no objection to people adding stuff here in the future.

For now I'm going to merge this.

**Pull request merged**